File: a5/other/q2Philosopher.java
public class q2Philosopher extends Thread {
	
	private int id, numberOfNoodles;
	private q2table table;
	private q2Printer printer;
    
    public enum States {
		Thinking( 'T' ), Hungry( 'H' ), Eating( 'E' ), Waiting( 'W' ), Finished( 'F' );
		public final char val;
		States( char v ) { val = v; }
    } // Philosopher.states
    
    private void doYield(int amount){
		for(int i=0; i<amount; i++){
			yield();
			   		try { sleep(1);
			}catch(InterruptedException ex){
				System.out.println("ERROR " + ex);
			}
			
		}
	}
		
	public q2Philosopher(int id, int noodles, q2table table, q2Printer prt){
		this.id = id;
		numberOfNoodles = noodles;
		this.table = table;
		printer = prt;	
	}	
	
	//start thread
	public void run(){

		while(true){
			printer.print(id, States.Hungry);
			
			//yield
			doYield(q2PRNG.getRand(0,4));
			
			//pickup forks from table
			table.pickup(id);
			
			//calc amount to eat
			int eatAmount = (q2PRNG.getRand(1,5));
			
			//upper bound
			if(eatAmount > numberOfNoodles)
				eatAmount = numberOfNoodles;
			
			numberOfNoodles -= eatAmount;
			
			printer.print(id, States.Eating, eatAmount, numberOfNoodles);
			
			//finish condition
			if(numberOfNoodles == 0) {
				table.putdown(id);
				break;
			}
					
			doYield(q2PRNG.getRand(0,4));		
			
			//putdown forks
			table.putdown(id);
			
			printer.print(id, States.Thinking);
			
			doYield(q2PRNG.getRand(0,19));
		}
		
		printer.print(id, States.Finished);
	}



}

File: a5/other/q2Philosopher.cc
#include <uC++.h>
#include "q2Philosopher.h"
//#include "q2table.h"
#include "q2Printer.h"
#include "q2PRNG.h"

extern PRNG myPrng;

void Philosopher::main(){

	while(true){
	
		printer->print(id, Philosopher::Hungry);
		
		//wait
		yield(myPrng(4));
		
		//grab fork
		table->pickup(id);
		
		//calc amount to eat
		unsigned int eatAmount = myPrng(1,5);
		
		//upper bound
		if(eatAmount > numberOfNoodles)
			eatAmount = numberOfNoodles;
		
		assert(eatAmount < 6);
		numberOfNoodles -= eatAmount;
		
		printer->print(id, Philosopher::Eating, eatAmount, numberOfNoodles);
		
		
		if(numberOfNoodles == 0) {
			table->putdown(id);
			break;
		}
				
		yield(myPrng(4));
		
		//put down fork
		table->putdown(id);
			
		printer->print(id, Philosopher::Thinking);
		
		yield(myPrng(19));
		
	}
	
	//finish message
	printer->print(id, Philosopher::Finished);
	
}

Philosopher::Philosopher( unsigned int id, unsigned int noodles, Table &table, Printer &prt ):id(id),numberOfNoodles(noodles),table(&table),printer(&prt){}
